pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,1145 @@
+                );
+                let wal_dir = self
+                    .conf
+                    .timeline_path(&timelineid, &self.tenantid)
+                    .join("wal");
+                import_timeline_wal(&wal_dir, &timeline, timeline.get_last_record_lsn())?;
+
+                let timeline_rc = Arc::new(timeline);
+                timelines.insert(timelineid, timeline_rc.clone());
+                Ok(timeline_rc)
+            }
+        }
+    }
+
+    pub fn new(
+        conf: &'static PageServerConf,
+        walredo_mgr: Arc<dyn WalRedoManager + Send + Sync>,
+        tenantid: ZTenantId,
+    ) -> LayeredRepository {
+        LayeredRepository {
+            tenantid: tenantid,
+            conf: conf,
+            timelines: Mutex::new(HashMap::new()),
+            walredo_mgr,
+        }
+    }
+
+    ///
+    /// Launch the checkpointer thread in given repository.
+    ///
+    pub fn launch_checkpointer_thread(conf: &'static PageServerConf, rc: Arc<LayeredRepository>) {
+        let _gc_thread = std::thread::Builder::new()
+            .name("Checkpointer thread".into())
+            .spawn(move || {
+                // FIXME: relaunch it? Panic is not good.
+                rc.checkpoint_loop(conf).expect("Checkpointer thread died");
+            })
+            .unwrap();
+    }
+
+    ///
+    /// Checkpointer thread's main loop
+    ///
+    fn checkpoint_loop(&self, conf: &'static PageServerConf) -> Result<()> {
+        loop {
+            std::thread::sleep(conf.gc_period);
+
+            info!("checkpointer thread for tenant {} waking up", self.tenantid);
+
+            // checkpoint timelines that have accumulated more than CHECKPOINT_INTERVAL
+            // bytes of WAL since last checkpoint.
+            {
+                let timelines = self.timelines.lock().unwrap();
+                for (_timelineid, timeline) in timelines.iter() {
+                    let distance = u64::from(timeline.last_valid_lsn.load())
+                        - u64::from(timeline.last_checkpoint_lsn.load());
+                    if distance > CHECKPOINT_INTERVAL {
+                        timeline.checkpoint()?;
+                    }
+                }
+                // release lock on 'timelines'
+            }
+
+            // Garbage collect old files that are not needed for PITR anymore
+            if conf.gc_horizon > 0 {
+                self.gc_iteration(None, conf.gc_horizon, false).unwrap();
+            }
+        }
+    }
+
+    /// Save timeline metadata to file
+    fn save_metadata(
+        conf: &'static PageServerConf,
+        timelineid: ZTimelineId,
+        tenantid: ZTenantId,
+        data: &TimelineMetadata,
+    ) -> Result<()> {
+        let path = conf.timeline_path(&timelineid, &tenantid).join("metadata");
+        let mut file = File::create(&path)?;
+
+        info!("saving metadata {}", path.display());
+
+        file.write_all(&TimelineMetadata::ser(data)?)?;
```

Comment:
Don't we need to call  `file.flush()?;` ?

Reply:
Opened https://github.com/zenithdb/zenith/issues/427 to track this